### PR TITLE
Update shim dependencies

### PR DESCRIPTION
**Pull request by mxpv**
This PR updates shim dependencies (updates ttrpc to latest versions).
Since TTRPC is now using latest protobuf 3.1, there is some migration work.

Close https://github.com/containerd/rust-extensions/pull/93 https://github.com/containerd/rust-extensions/pull/92 https://github.com/containerd/rust-extensions/pull/91 https://github.com/containerd/rust-extensions/pull/82

### CODE REVIEW

File: crates/runc-shim/src/synchronous/container.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@ use std::os::unix::io::{AsRawFd, FromRawFd};
 use std::path::Path;
 use std::sync::mpsc::{sync_channel, Receiver, SyncSender};
 
+use containerd_shim::protos::protobuf::EnumOrUnknown;
```
Comment (Burning1020) [inline crates/runc-shim/src/synchronous/container.rs:24]: should we make a rule of referencing packages order, like std/3rd/self

File: crates/shim-protos/src/shim/mod.rs [hunk 2]
```diff
@@ -13,3 +13,6 @@ pub mod shim_ttrpc_async;
 pub(crate) mod empty;
 pub(crate) mod mount;
 pub(crate) mod task;
+
+mod fieldpath;
+mod gogo;
```
Comment (Burning1020) [inline crates/shim-protos/src/shim/mod.rs:18]: public or not?

File: crates/shim/src/cgroup.rs [hunk 3]
```diff
@@ -109,19 +109,19 @@ pub fn collect_metrics(pid: u32) -> Result<Metrics> {
         match sub_system {
             Subsystem::CpuAcct(cpuacct_ctr) => {
                 let mut cpu_usage = CPUUsage::new();
-                cpu_usage.set_total(cpuacct_ctr.cpuacct().usage);
+                cpu_usage.total = cpuacct_ctr.cpuacct().usage;
                 let mut cpu_stat = CPUStat::new();
-                cpu_stat.set_usage(cpu_usage);
-                metrics.set_cpu(cpu_stat);
+                cpu_stat.usage = Some(cpu_usage).into();
```
Comment (Burning1020) [inline crates/shim/src/cgroup.rs:114]: Does the new ttrpc not provide setter functions? I think `Some(xxx).into` is a little complex by writing.

Comment (mxpv) [reply crates/shim/src/cgroup.rs:114]: Protobuf 3 has removed setters/getters for generated structs.
`usage` has [MessageField](https://docs.rs/protobuf/latest/protobuf/struct.MessageField.html) type, which is a wrapper for `Option<Box<T>>` (replacement for `SingularPtrField` used in Protobuf 2). I found `Some().into()` to be the shortest syntax possible here.

File: crates/shim-protos/src/shim/mod.rs (hunk 2 above)
Comment (mxpv) [reply crates/shim-protos/src/shim/mod.rs:18]: I plan to remove the generated code from the repo, so I don't have strong preference here.

File: crates/runc-shim/src/synchronous/container.rs (hunk 1 above)
Comment (mxpv) [reply crates/runc-shim/src/synchronous/container.rs:25]: I've fixed imports in https://github.com/containerd/rust-extensions/pull/95/commits/702d0212f4bf398c17a1d70c5bda28577fb19475
Can you elaborate on "make a rule"? I could not find the way to enforce imports order on CI :(

File: crates/shim/src/cgroup.rs (hunk 3 above)
Comment (mxpv) [reply crates/shim/src/cgroup.rs:114]: It's also possible to use `MessageField::some()`, but it looks weird in the context of Rust with std `Option` type. So I'd prefer to use `Some` here.

Comment (Burning1020) [reply crates/shim/src/cgroup.rs:114]: that's OK

File: crates/runc-shim/src/synchronous/container.rs (hunk 1 above)
Comment (Burning1020) [reply crates/runc-shim/src/synchronous/container.rs:25]: Same with golang projects, I need to seperate these packages into three parts(std/3rd/self) by myself, the order of each part will be updated automatically by IDE or rust plugins. Thus, is it necessary to make demarcation? I also see many other repos they don't care about that.

File: crates/runc-shim/src/synchronous/container.rs [hunk 4]
```diff
@@ -35,6 +34,7 @@ use shim::protos::cgroups::metrics::Metrics;
 use shim::util::{convert_to_timestamp, read_pid_from_file};
 use shim::Console;
 use shim::{io_error, other, other_error};
+use time::OffsetDateTime;
```
Comment (Burning1020) [inline crates/runc-shim/src/synchronous/container.rs:37]: mark

File: crates/runc-shim/src/synchronous/container.rs [hunk 5]
```diff
@@ -21,11 +21,10 @@ use std::os::unix::io::{AsRawFd, FromRawFd};
 use std::path::Path;
 use std::sync::mpsc::{sync_channel, Receiver, SyncSender};
 
+use containerd_shim as shim;
+use containerd_shim::protos::protobuf::EnumOrUnknown;
```
Comment (Burning1020) [inline crates/runc-shim/src/synchronous/container.rs:25]: mark

File: crates/runc-shim/src/synchronous/service.rs [hunk 6]
```diff
@@ -21,6 +21,7 @@ use std::path::Path;
 use std::sync::mpsc::{channel, Receiver, Sender};
 use std::sync::Arc;
 
+use containerd_shim::protos::protobuf::MessageDyn;
```
Comment (Burning1020) [inline crates/runc-shim/src/synchronous/service.rs:24]: mark

File: crates/shim/src/asynchronous/processes.rs [hunk 7]
```diff
@@ -18,13 +18,14 @@ use std::os::unix::io::AsRawFd;
 use std::sync::Arc;
 
 use async_trait::async_trait;
+use containerd_shim_protos::protobuf::EnumOrUnknown;
```
Comment (Burning1020) [inline crates/shim/src/asynchronous/processes.rs:21]: mark

File: crates/runc-shim/src/synchronous/container.rs (hunk 1 above)
Comment (Burning1020) [reply crates/runc-shim/src/synchronous/container.rs:25]: If it's troublesome, just let it go as it doesn't impact compile.

**Comment by mxpv** [comment]
So I just found `generate_accessors` toggle that generated accessors. So I've reverted some code and got rid of these extra Enum imports. Unfortunately this does not make the diff a lot less, because accessors were renamed from `get_xyz()` to `xyz()`

**Review by fuweid — APPROVED** [review]

**Comment by fuweid** [comment]
ping @Burning1020

**Comment by Burning1020** [comment]
Lgtm

**Pull request merged**